libp2p/protocols/secure/secio.nim

```diff
@@ -352,41 +354,40 @@ proc handshake*(s: Secio, conn: Connection): Future[SecureConnection] {.async.}
   var signature = s.localPrivateKey.sign(localCorpus)
 
   var localExchange = createExchange(epubkey, signature.getBytes())
-
   var remoteExchange = await transactMessage(conn, localExchange)
   if len(remoteExchange) == 0:
     trace "Corpus exchange failed", conn = conn
-    return
+    raise newException(SecioError, "Corpus exchange failed")
```

Comment:
wouldn't this be the place to close conn already? or can it still be salvaged?

Reply:
ie you don't know what's coming next on the wire in this case, right?

Reply:
This would immediately hit the wrapping `finally` and close the connection there, otherwise you'd be repeating `conn.close()` before each raise. DRY.

Reply:
the wrapping finally is a bit late though - why should the caller of this api have to take extra actions on exceptions? they're not the one that understand what's going on on the wire and whether the connection can be salvaged or not since they're not doing the messaging. crucially, it is known that the connection is broken _before_ the raise and certainly before the try/finally - else you have to document it as part of the api that every time you call this funciton you have to wrap and close on (some?) exceptions.

also, there's nothing forcing the caller to close - they have to explicitly remember it - this is a source of bugs because _inevitably_ they will get it wrong - it's not DRY when every user of the API has to reimplement this error handling.. by closing here, you prevent them from making this mistake and others (like trying to send something else on the connection)..

Reply:
This isn't public API (even tho `handshake` is exposed with `*`, it shouldn't be), this is an internal method that isn't supposed to be called outside the module. All relevant error handling happens within the module, so the consumer of the API shouldn't have to worry about doing any sort of exception/error handling. If the connection could not be secured, then it's properly cleaned up in the `finally` and the caller is notified by way of a `nil` return value.

Reply:
even for private api, consider that the shorter the inconsistent `conn` state exists, the smaller the surface area of potential bugs and confusing behaviours around it - how it breaks and how you react to it is "local" and you don't need to reverse-look-for-callers between pieces of code to find out what's going on

Reply:
But isn't `finally` "atomic"? There shouldn't be any inconsistent state between when an exception happens and when `finally` hits - honest question, this is my assumption coming to Nim from other languages that do guarantee this. Of Course, in a multithreaded environment it gets a bit trickier, but this is a general problem with concurrency and locking, not just error handling.

> how it breaks and how you react to it is "local" and you don't need to reverse-look-for-callers between pieces of code to find out what's going on

I actually disagree, error handling should be as generic and simple as possible, otherwise it's going to be a source of errors/bugs in and out of itself. I prefer handling errors and release resources in one place rather than split/duplicate it in a bunch of sites; and exceptions with `try/finally` encourages this and it is why I prefer them over return values in general. I agree they have their own shortcomings and currently the pendulum is swinging back to return values over exceptions (in fact there is a general tendency towards minimalistic languages, which I overall wellcome), but this doesn't change the premise of simple and generic error handling, hopefully localized to one or a small number of sites. 

To me, what you're suggesting (closing the `conn` before throwing) feels like breaking that premise of generic and localized error handling.

Reply:
well, the sequence is: fail, raise, finally - 3 steps.. not atomic - after you get the fail notification, conn is broken, and then you rely on someone else remembering to reset it, from the local point of view of the code that detects the failure. that's what I mean by local - in the same function - else, between the finally handler and the failure point, you open up a surface area for misuse of conn, that will show its ugly face during refactoring or panic:ed hot-bug-fixing.. "can't go wrong" vs "shouldn't go wrong unless programmer forgets" - same argument as we had initially with the nil pointer - "of course a programmer wouldn't write code with both nil and exceptions!"... "oops"

fwiw, c++ typically deals with this by using raii and similar tricks, so that you don't have to manually deal with it. in nim, you have to structure things a bit different (or use even simpler code)

Reply:
> well, the sequence is: fail, raise, finally - 3 steps.. not atomic 

I meant atomic from the standpoint of introducing more instructions by the programmer, once an exception is raised, you're done and `finally` hits right afterwards. Things get murkier when dealing with concurrency, but at that point, you're already have to take care of proper locking and synchronization, so no overhead there either.

> after you get the fail notification, conn is broken, and then you rely on someone else remembering to reset it, from the local point of view of the code that detects the failure. that's what I mean by local - in the same function - else, between the finally handler and the failure point, you open up a surface area for misuse of conn, that will show its ugly face during refactoring or panic:ed hot-bug-fixing.. "can't go wrong" vs "shouldn't go wrong unless programmer forgets" - same argument as we had initially with the nil pointer - "of course a programmer wouldn't write code with both nil and exceptions!"... "oops"

This is still pretty subjective, the same argument can be used against it: now you have to remember all the sites where `conn.close()` has to be changed during refactoring. What's worst, having it in one specific site or spread across many sites? IMO, the smaller the surface of changes during refactoring the lower the risk of introducing new errors. 

> fwiw, c++ typically deals with this by using raii and similar tricks, so that you don't have to manually deal with it. in nim, you have to structure things a bit different (or use even simpler code)

Yes, this would be a welcome addition to Nim, but it isn't much different form what `finally` already does for you in practice. Sure, RAII has better guarantees, but the argument for not introducing RAII in GC languages is usually the lack of control over the lifecycle of the object, so we go with `finally` instead.

I think it would be helpful to take this discussion further, I'm glad this is happening because it allows us to re-evaluate many of the assumptions that we've come to take for granted - which is always a good thing. However, a lot of the things you're bringing up become controversial for at least two reasons (strictly IMO, others might disagree):

a) They seem counterintuitive in many cases and I think this discussion is a good example of that. For example, I'm relying on well accepted practices about code duplication and proper usage of language constructs (like `try/finally`) for resource management/cleanup. The arguments you bring up to question this practices are less widespread and seem to be more anecdotal (please correct me if I'm wrong). Now this isn't a bad thing, as I said, re-evaluating one's assumptions is always good, but to accept them as "better" than the status quo, one would need a lot more evidence.
b) Following from the above, accepting this claims/suggestions are a major investment in terms of changing one's coding style and I would suspect it's an investment that not everyone is willing to take on unless it has been properly augmented and justified.

With that, I propose we create a metaissue where this "best practices" get discussed and everyone chips in with proper research and data along their own reasoning for preferring one style over another.

So far I've seen two major issues pop up (both related): 

- Error handling (exceptions vs return values)
- Resource management/object initialization